File: dev/test_data/main/golang/expected/aas_core_meta.v3/input/snippets/Verification/value_consistent_with_XSD_type.go
var zero *big.Int = big.NewInt(0)
var one *big.Int = big.NewInt(1)
var four *big.Int = big.NewInt(4)
var hundred *big.Int = big.NewInt(100)
var fourHundred *big.Int = big.NewInt(400)

// Check if `year` is a leap year.
func isLeapYear(year *big.Int) bool {
	// We consider the years B.C. to be one-off.
	//
	// See the note at: https://www.w3.org/TR/xmlschema-2/#dateTime:
	// "'-0001' is the lexical representation of the year 1 Before Common Era
	// (1 BCE, sometimes written "1 BC")."
	//
	// Hence, -1 year in XML is 1 BCE, which is 0 year in astronomical years.

	if year.Cmp(zero) < 0 {
		// year = -year - 1
		year.Neg(year)
		year.Sub(year, one)
	}

	// See: https://en.wikipedia.org/wiki/Leap_year#Algorithm
	yearMod4 := &big.Int{}
	yearMod4.Mod(year, four)

	if yearMod4.Cmp(zero) != 0 {
		return false
	}

	yearMod100 := &big.Int{}
	yearMod100.Mod(year, hundred)
	if yearMod100.Cmp(zero) != 0 {
		return true
	}

	yearMod400 := &big.Int{}
	yearMod400.Mod(year, fourHundred)
	if yearMod400.Cmp(zero) != 0 {
		return false
	}

	return true
}

var daysInMonth = []int{
	// Month 0 is not defined.
	-1, // 0
	31, // 1
	// Please use isLeapYear if you need to check whether
	// a concrete February has 28 or 29 days.
	29, // 2
	31, // 3
	30, // 4
	31, // 5
	30, // 6
	31, // 7
	31, // 8
	30, // 9
	31, // 10
	30, // 11
	31, // 12
}

var datePrefixRe = regexp.MustCompile("^(-?[0-9]+)-([0-9]{2})-([0-9]{2})")

// Check that `value` is a valid `xs:date`.
//
// The year must fit in the 64-bit range so that we can check whether it is
// a leap year or not.
func IsXsDate(value string) bool {
	if !MatchesXsDate(value) {
		return false
	}

	// NOTE (mristin, 2023-05-10):
	// We can not use the date functions from the standard library as we have
	// to handle years BCE according to the XML date type.

	// NOTE (mristin, 2023-05-12):
    // We need to match the prefix as zone offsets are allowed in the dates. Optimally,
    // we would re-use the pattern matching from `MatchesXsDate`, but this
    // would make the code generation and constraint inference for schemas much more
    // difficult. Hence, we sacrifice the efficiency a bit for the clearer code & code
    // generation.

	match := datePrefixRe.FindStringSubmatch(value)
	if len(match) == 0 {
		panic(
			fmt.Sprintf(
				"Expected value to match %v if we got thus far, " +
				"but it does not: %s",
				datePrefixRe, value,
			),
		)
	}

	yearStr := match[1]
	monthStr := match[2]
	dayStr := match[3]

	year := &big.Int{}
	_, ok := year.SetString(yearStr, 10)
	if !ok {
		panic(
			fmt.Sprintf(
				"Failed to convert the year from %s",
				yearStr,
			),
		)
	}

	month, err := strconv.Atoi(monthStr)
	if err != nil {
		panic(
			fmt.Sprintf(
				"Failed to convert the month from %s: %s",
				monthStr, err.Error(),
			),
		)
	}

	var day int
	day, err = strconv.Atoi(dayStr)
	if err != nil {
		panic(
			fmt.Sprintf(
				"Failed to convert the day from %s: %s",
				dayStr, err.Error(),
			),
		)
	}

	// We do not accept year zero,
	// see the note at: https://www.w3.org/TR/xmlschema-2/#dateTime
	if year.Cmp(zero) == 0 {
		return false
	}

	if day <= 0 {
		return false
	}

	if month <= 0 || month >= 13 {
		return false
	}

	var maxDays int
	if month == 2 {
		if isLeapYear(year) {
			maxDays = 29
		} else {
			maxDays = 28
		}
	} else {
		maxDays = daysInMonth[month]
	}

	if day > maxDays {
		return false
	}

	return true
}

// Check that `value` is a valid `xs:double`.
func IsXsDouble(value string) bool {
	// We need to check explicitly for the regular expression since
	// strconv.ParseFloat is too permissive. For example, it accepts "nan"
	// although only "NaN" is valid.
	// See: https://www.w3.org/TR/xmlschema-2/#double
	if !MatchesXsDouble(value) {
		return false
	}

	_, err := strconv.ParseFloat(value, 64)
	if err != nil {
		if numError, ok := err.(*strconv.NumError); ok {
            if numError.Err == strconv.ErrRange {
                return false
            }
        }

		panic(
			fmt.Sprintf(
				"Failed to parse float from value %s: %s",
				value, err.Error(),
			),
		)
	}

	// NOTE (2023-05-12):
	// We explicitly do not check for loss of precision, as the majority of people will
	// use string representation of the floating point numbers ignoring the precision
	// issues. For example, the closest double-precision number to the number `359.9` is
	// `359.8999999999999772626324556767940521240234375`, but most people will simply
	// give `359.9` as the value.

	return true
}

// Check that `value` is a valid `xs:float`.
func IsXsFloat(value string) bool {
	// We need to check explicitly for the regular expression since
	// strconv.ParseFloat is too permissive. For example, it accepts "nan"
	// although only "NaN" is valid.
	// See: https://www.w3.org/TR/xmlschema-2/#double
	if !MatchesXsDouble(value) {
		return false
	}

	_, err := strconv.ParseFloat(value, 32)
	if err != nil {
        if numError, ok := err.(*strconv.NumError); ok {
            if numError.Err == strconv.ErrRange {
                return false
            }
        }

		panic(
			fmt.Sprintf(
				"Failed to parse float from value %s: %s",
				value, err.Error(),
			),
		)
	}

	// NOTE (2023-05-12):
	// We explicitly do not check for loss of precision, as the majority of people will
	// use string representation of the floating point numbers ignoring the precision
	// issues. For example, `float64(float32(3.2)) == 3.2` is false in Golang, but "3.2"
	// is totally expected as a value.

	return true
}

// Check that `value` is a valid `xs:gMonthDay`.
func IsXsGMonthDay(value string) bool {
	if !MatchesXsGMonthDay(value) {
		return false
	}

	month, err := strconv.Atoi(value[2:4])
	if err != nil {
		panic(
			fmt.Sprintf(
				"Unexpected fail to parse the month %s: %s",
				value[2:4], err.Error(),
			),
		)
	}

	var day int
	day, err = strconv.Atoi(value[5:7])
	if err != nil {
		panic(
			fmt.Sprintf(
				"Unexpected fail to parse the day %s: %s",
				value[5:7], err.Error(),
			),
		)
	}

	maxDays := daysInMonth[month]
	return day <= maxDays
}

// Check that `value` is a valid `xs:long`.
func IsXsLong(value string) bool {
	if !MatchesXsLong(value) {
		return false
	}

	_, err := strconv.ParseInt(value, 10, 64)
	return err == nil
}

// Check that `value` is a valid `xs:int`.
func IsXsInt(value string) bool {
	if !MatchesXsInt(value) {
		return false
	}

	_, err := strconv.ParseInt(value, 10, 32)
	return err == nil
}

// Check that `value` is a valid `xs:short`.
func IsXsShort(value string) bool {
	if !MatchesXsShort(value) {
		return false
	}

	_, err := strconv.ParseInt(value, 10, 16)
	return err == nil
}

// Check that `value` is a valid `xs:byte`.
func IsXsByte(value string) bool {
	if !MatchesXsByte(value) {
		return false
	}

	_, err := strconv.ParseInt(value, 10, 8)
	return err == nil
}

// Check that `value` is a valid `xs:unsignedLong`.
func IsXsUnsignedLong(value string) bool {
	if !MatchesXsUnsignedLong(value) {
		return false
	}

	// See: https://pkg.go.dev/strconv#ParseUint,
	// "A sign prefix is not permitted."
	if value[0] == '+' {
		value = value[1:]
	}

	_, err := strconv.ParseUint(value, 10, 64)
	return err == nil
}

// Check that `value` is a valid `xs:unsignedInt`.
func IsXsUnsignedInt(value string) bool {
	if !MatchesXsUnsignedInt(value) {
		return false
	}

	// See: https://pkg.go.dev/strconv#ParseUint,
	// "A sign prefix is not permitted."
	if value[0] == '+' {
		value = value[1:]
	}

	_, err := strconv.ParseUint(value, 10, 32)
	return err == nil
}

// Check that `value` is a valid `xs:unsignedShort`.
func IsXsUnsignedShort(value string) bool {
	if !MatchesXsUnsignedShort(value) {
		return false
	}

	// See: https://pkg.go.dev/strconv#ParseUint,
	// "A sign prefix is not permitted."
	if value[0] == '+' {
		value = value[1:]
	}

	_, err := strconv.ParseUint(value, 10, 16)
	return err == nil
}

// Check that `value` is a valid `xs:unsignedByte`.
func IsXsUnsignedByte(value string) bool {
	if !MatchesXsUnsignedByte(value) {
		return false
	}

	// See: https://pkg.go.dev/strconv#ParseUint,
	// "A sign prefix is not permitted."
	if value[0] == '+' {
		value = value[1:]
	}

	_, err := strconv.ParseUint(value, 10, 8)
	return err == nil
}

// Check that `value` is consistent with the given `valueType`.
func ValueConsistentWithXSDType(
	value string,
	valueType aastypes.DataTypeDefXSD,
) bool {
	switch valueType {
		case aastypes.DataTypeDefXSDAnyURI:
			return MatchesXsAnyURI(value)
		case aastypes.DataTypeDefXSDBase64Binary:
			return MatchesXsBase64Binary(value)
		case aastypes.DataTypeDefXSDBoolean:
			return MatchesXsBoolean(value)
		case aastypes.DataTypeDefXSDByte:
			return IsXsByte(value)
		case aastypes.DataTypeDefXSDDate:
			return IsXsDate(value)
		case aastypes.DataTypeDefXSDDateTime:
			return IsXsDateTime(value)
		case aastypes.DataTypeDefXSDDecimal:
			return MatchesXsDecimal(value)
		case aastypes.DataTypeDefXSDDouble:
			return IsXsDouble(value)
		case aastypes.DataTypeDefXSDDuration:
			return MatchesXsDuration(value)
		case aastypes.DataTypeDefXSDFloat:
		 	return IsXsFloat(value)
		case aastypes.DataTypeDefXSDGDay:
			return MatchesXsGDay(value)
		case aastypes.DataTypeDefXSDGMonth:
			return MatchesXsGMonth(value)
		case aastypes.DataTypeDefXSDGMonthDay:
			return IsXsGMonthDay(value)
		case aastypes.DataTypeDefXSDGYear:
			return MatchesXsGYear(value)
		case aastypes.DataTypeDefXSDGYearMonth:
			return MatchesXsGYearMonth(value)
		case aastypes.DataTypeDefXSDHexBinary:
			return MatchesXsHexBinary(value)
		case aastypes.DataTypeDefXSDInt:
			return IsXsInt(value)
		case aastypes.DataTypeDefXSDInteger:
			return MatchesXsInteger(value)
		case aastypes.DataTypeDefXSDLong:
			return IsXsLong(value)
		case aastypes.DataTypeDefXSDNegativeInteger:
			return MatchesXsNegativeInteger(value)
		case aastypes.DataTypeDefXSDNonNegativeInteger:
			return MatchesXsNonNegativeInteger(value)
		case aastypes.DataTypeDefXSDNonPositiveInteger:
			return MatchesXsNonPositiveInteger(value)
		case aastypes.DataTypeDefXSDPositiveInteger:
			return MatchesXsPositiveInteger(value)
		case aastypes.DataTypeDefXSDShort:
			return IsXsShort(value)
		case aastypes.DataTypeDefXSDString:
			return MatchesXsString(value)
		case aastypes.DataTypeDefXSDTime:
			return MatchesXsTime(value)
		case aastypes.DataTypeDefXSDUnsignedByte:
			return IsXsUnsignedByte(value)
		case aastypes.DataTypeDefXSDUnsignedInt:
			return IsXsUnsignedInt(value)
		case aastypes.DataTypeDefXSDUnsignedLong:
			return IsXsUnsignedLong(value)
		case aastypes.DataTypeDefXSDUnsignedShort:
			return IsXsUnsignedShort(value)
		default:
			panic(fmt.Sprintf("Unhandled value type: %v", valueType))
	}
}

File: dev/test_data/main/python/expected/aas_core_meta.v3/input/snippets/Verification/value_consistent_with_XSD_type.py
def _is_leap_year(year: int) -> bool:
    """
    Check if :paramref:`year` is a leap year.

    >>> _is_leap_year(2016)
    True

    >>> _is_leap_year(1700)
    False

    >>> _is_leap_year(1600)
    True

    >>> _is_leap_year(2000)
    True
    """
    # We consider the years B.C. to be one-off.
    #
    # See the note at: https://www.w3.org/TR/xmlschema-2/#dateTime:
    # "'-0001' is the lexical representation of the year 1 Before Common Era
    # (1 BCE, sometimes written "1 BC")."
    #
    # Hence, -1 year in XML is 1 BCE, which is 0 year in astronomical years.
    if year < 0:
        year = abs(year) - 1

    # See: https://en.wikipedia.org/wiki/Leap_year#Algorithm
    if year % 4 > 0:
        return False

    if year % 100 > 0:
        return True

    if year % 400 > 0:
        return False

    return True


_DAYS_IN_MONTH: Mapping[int, int] = {
    1: 31,
    # Please use _is_leap_year if you need to check
    # whether a concrete February has 28 or 29 days.
    2: 29,
    3: 31,
    4: 30,
    5: 31,
    6: 30,
    7: 31,
    8: 31,
    9: 30,
    10: 31,
    11: 30,
    12: 31,
}


_DATE_PREFIX_RE = re.compile(r"^(-?[0-9]+)-([0-9]{2})-([0-9]{2})")


def is_xs_date(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:date``."""
    if not matches_xs_date(value):
        return False

    # NOTE (mristin, 2022-11-23):
    # We can not use :py:func:`datetime.datetime.strptime` as it does not
    # handle years below 1000 correctly on Windows (*e.g.*, ``-999-01-01``).

    # NOTE (mristin, 2022-10-30):
    # We need to match the prefix as zone offsets are allowed in the dates. Optimally,
    # we would re-use the pattern matching from :py:func`matches_xs_date`, but this
    # would make the code generation and constraint inference for schemas much more
    # difficult. Hence, we sacrifice the efficiency a bit for the clearer code & code
    # generation.
    match = _DATE_PREFIX_RE.match(value)
    assert match is not None

    year = int(match.group(1))
    month = int(match.group(2))
    day = int(match.group(3))

    # We do not accept year zero,
    # see the note at: https://www.w3.org/TR/xmlschema-2/#dateTime
    if year == 0:
        return False

    if day <= 0:
        return False

    if month <= 0 or month >= 13:
        return False

    if month == 2:
        max_days = 29 if _is_leap_year(year) else 28
    else:
        max_days = _DAYS_IN_MONTH[month]

    if day > max_days:
        return False

    return True


def is_xs_double(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:double``."""
    # We need to check explicitly for the regular expression since
    # ``float(.)`` is too permissive. For example,
    # it accepts "nan" although only "NaN" is valid.
    # See: https://www.w3.org/TR/xmlschema-2/#double
    if not matches_xs_double(value):
        return False

    converted = float(value)

    # Check that the value is either "INF" or "-INF".
    # Otherwise, the value is a decimal which is too big
    # to be represented as a double-precision floating point
    # number.
    #
    # Python simply rounds up/down to ``INF`` and ``-INF``,
    # respectively, if the number is too large.
    # For example: ``float("1e400") == math.inf``
    if math.isinf(converted) and value != "INF" and value != "-INF":
        return False

    return True


def is_xs_float(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:float``."""
    # We need to check explicitly for the regular expression since
    # ``float(.)`` is too permissive. For example,
    # it accepts "nan" although only "NaN" is valid.
    # See: https://www.w3.org/TR/xmlschema-2/#double
    if not matches_xs_float(value):
        return False

    converted = float(value)

    # Check that the value is either "INF" or "-INF".
    # Otherwise, the value is a decimal which is too big
    # to be represented as a single-precision floating point
    # number.
    #
    # Python simply rounds up/down to ``INF`` and ``-INF``,
    # respectively, if the number is too large.
    # For example: ``float("1e400") == math.inf``
    if math.isinf(converted) and value != "INF" and value != "-INF":
        return False

    # Python uses double-precision floating point numbers. Since
    # we check for a single-precision one, we have to explicitly
    # see if the number is within a range of a single-precision
    # floating point numbers.
    try:
        _ = struct.pack(">f", converted)
    except OverflowError:
        return False

    return True


def is_xs_g_month_day(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:gMonthDay``."""
    if not matches_xs_g_month_day(value):
        return False

    month = int(value[2:4])
    day = int(value[5:7])

    max_days = _DAYS_IN_MONTH[month]
    return day <= max_days


def is_xs_long(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:long``."""
    if not matches_xs_long(value):
        return False

    converted = int(value)
    return -9223372036854775808 <= converted <= 9223372036854775807


def is_xs_int(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:int``."""
    if not matches_xs_int(value):
        return False

    converted = int(value)
    return -2147483648 <= converted <= 2147483647


def is_xs_short(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:short``."""
    if not matches_xs_short(value):
        return False

    converted = int(value)
    return -32768 <= converted <= 32767


def is_xs_byte(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:byte``."""
    if not matches_xs_byte(value):
        return False

    converted = int(value)
    return -128 <= converted <= 127


def is_xs_unsigned_long(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:unsignedLong``."""
    if not matches_xs_unsigned_long(value):
        return False

    converted = int(value)
    return 0 <= converted <= 18446744073709551615


def is_xs_unsigned_int(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:unsignedInt``."""
    if not matches_xs_unsigned_int(value):
        return False

    converted = int(value)
    return 0 <= converted <= 4294967295


def is_xs_unsigned_short(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:unsignedShort``."""
    if not matches_xs_unsigned_short(value):
        return False

    converted = int(value)
    return 0 <= converted <= 65535


def is_xs_unsigned_byte(value: str) -> bool:
    """Check that :paramref:`value` is a valid ``xs:unsignedByte``."""
    if not matches_xs_unsigned_byte(value):
        return False

    converted = int(value)
    return 0 <= converted <= 255


_DATA_TYPE_DEF_XSD_TO_VALUE_CONSISTENCY: Mapping[
    aas_types.DataTypeDefXSD, Callable[[str], bool]
] = {
    aas_types.DataTypeDefXSD.ANY_URI: matches_xs_any_uri,
    aas_types.DataTypeDefXSD.BASE_64_BINARY: matches_xs_base_64_binary,
    aas_types.DataTypeDefXSD.BOOLEAN: matches_xs_boolean,
    aas_types.DataTypeDefXSD.BYTE: is_xs_byte,
    aas_types.DataTypeDefXSD.DATE: is_xs_date,
    aas_types.DataTypeDefXSD.DATE_TIME: is_xs_date_time,
    aas_types.DataTypeDefXSD.DECIMAL: matches_xs_decimal,
    aas_types.DataTypeDefXSD.DOUBLE: is_xs_double,
    aas_types.DataTypeDefXSD.DURATION: matches_xs_duration,
    aas_types.DataTypeDefXSD.FLOAT: is_xs_float,
    aas_types.DataTypeDefXSD.G_DAY: matches_xs_g_day,
    aas_types.DataTypeDefXSD.G_MONTH: matches_xs_g_month,
    aas_types.DataTypeDefXSD.G_MONTH_DAY: is_xs_g_month_day,
    aas_types.DataTypeDefXSD.G_YEAR: matches_xs_g_year,
    aas_types.DataTypeDefXSD.G_YEAR_MONTH: matches_xs_g_year_month,
    aas_types.DataTypeDefXSD.HEX_BINARY: matches_xs_hex_binary,
    aas_types.DataTypeDefXSD.INT: is_xs_int,
    aas_types.DataTypeDefXSD.INTEGER: matches_xs_integer,
    aas_types.DataTypeDefXSD.LONG: is_xs_long,
    aas_types.DataTypeDefXSD.NEGATIVE_INTEGER: matches_xs_negative_integer,
    aas_types.DataTypeDefXSD.NON_NEGATIVE_INTEGER: matches_xs_non_negative_integer,
    aas_types.DataTypeDefXSD.NON_POSITIVE_INTEGER: matches_xs_non_positive_integer,
    aas_types.DataTypeDefXSD.POSITIVE_INTEGER: matches_xs_positive_integer,
    aas_types.DataTypeDefXSD.SHORT: is_xs_short,
    aas_types.DataTypeDefXSD.STRING: matches_xs_string,
    aas_types.DataTypeDefXSD.TIME: matches_xs_time,
    aas_types.DataTypeDefXSD.UNSIGNED_BYTE: is_xs_unsigned_byte,
    aas_types.DataTypeDefXSD.UNSIGNED_INT: is_xs_unsigned_int,
    aas_types.DataTypeDefXSD.UNSIGNED_LONG: is_xs_unsigned_long,
    aas_types.DataTypeDefXSD.UNSIGNED_SHORT: is_xs_unsigned_short,
}
assert all(
    data_type_def_xsd in _DATA_TYPE_DEF_XSD_TO_VALUE_CONSISTENCY
    for data_type_def_xsd in aas_types.DataTypeDefXSD
)


def value_consistent_with_xsd_type(
    value: str, value_type: aas_types.DataTypeDefXSD
) -> bool:
    """
    Check that :paramref:`value` is consistent with the given
    :paramref:`value_type`.
    """
    return _DATA_TYPE_DEF_XSD_TO_VALUE_CONSISTENCY[value_type](value)
